Mathlib/RingTheory/Polynomial/Irreducible.lean

```diff
@@ -0,0 +1,146 @@
+  obtain ⟨k, rfl⟩ := hm
+  rintro rfl
+  obtain ⟨q, hq⟩ := sub_dvd_pow_sub_pow (X ^ k) (C b) m
+  rw [mul_comm, pow_mul, map_pow, hq] at H
+  have : degree q = 0
+  · simpa [isUnit_iff_degree_eq_zero, degree_X_pow_sub_C,
+      Nat.pos_iff_ne_zero, (mul_ne_zero_iff.mp hn).2] using H.2 _ q rfl
+  apply_fun degree at hq
+  simp only [this, ← pow_mul, mul_comm k m, degree_X_pow_sub_C, Nat.pos_iff_ne_zero.mpr hn,
+    Nat.pos_iff_ne_zero.mpr (mul_ne_zero_iff.mp hn).2, degree_mul, ← map_pow, add_zero,
+    Nat.cast_injective.eq_iff] at hq
+  exact hm' ((mul_eq_right₀ (mul_ne_zero_iff.mp hn).2).mp hq)
+
+theorem X_pow_sub_C_irreducible_of_prime {p : ℕ} (hp : p.Prime) {a : K} (ha : ∀ b : K, b ^ p ≠ a) :
+    Irreducible (X ^ p - C a) := by
+  -- First of all, We may find an irreducible factor `g` of `X ^ p - C a`.
+  have : ¬ IsUnit (X ^ p - C a)
+  · rw [Polynomial.isUnit_iff_degree_eq_zero, degree_X_pow_sub_C hp.pos, Nat.cast_eq_zero]
+    exact hp.ne_zero
+  have ⟨g, hg, hg'⟩ := WfDvdMonoid.exists_irreducible_factor this (X_pow_sub_C_ne_zero hp.pos a)
+  -- It suffices to show that `deg g = p`.
+  suffices natDegree g = p from (associated_of_dvd_of_natDegree_le hg'
+    (X_pow_sub_C_ne_zero hp.pos a) (this.trans natDegree_X_pow_sub_C.symm).ge).irreducible hg
+  -- Suppose `deg g ≠ p`.
+  by_contra h
+  have : Fact (Irreducible g) := ⟨hg⟩
+  -- Let `r` be a root of `g`, then `N_K(r) ^ p = N_K(r ^ p) = N_K(a) = a ^ (deg g)`.
+  have key : (Algebra.norm K (AdjoinRoot.root g)) ^ p = a ^ g.natDegree
+  · have := eval₂_eq_zero_of_dvd_of_eval₂_eq_zero _ _ hg' (AdjoinRoot.eval₂_root g)
+    rw [eval₂_sub, eval₂_pow, eval₂_C, eval₂_X, sub_eq_zero] at this
+    rw [← map_pow, this, ← AdjoinRoot.algebraMap_eq, Algebra.norm_algebraMap,
+      ← finrank_top', ← IntermediateField.adjoin_root_eq_top g,
+      IntermediateField.adjoin.finrank,
+      AdjoinRoot.minpoly_root hg.ne_zero, natDegree_mul_C]
+    · simpa using hg.ne_zero
+    · exact AdjoinRoot.isIntegral_root hg.ne_zero
+  -- Since `a ^ (deg g)` is a `p`-power, and `p` is coprime to `deg g`, we conclude that `a` is
+  -- also a `p`-power, contradicting the hypothesis
+  have : p.Coprime (natDegree g) := hp.coprime_iff_not_dvd.mpr (fun e ↦ h (((natDegree_le_of_dvd hg'
+    (X_pow_sub_C_ne_zero hp.pos a)).trans_eq natDegree_X_pow_sub_C).antisymm (Nat.le_of_dvd
+      (natDegree_pos_iff_degree_pos.mpr <| Polynomial.degree_pos_of_irreducible hg) e)))
+  exact ha _ ((pow_mem_range_pow_of_coprime this.symm a).mp ⟨_, key⟩).choose_spec
+
+theorem X_pow_sub_C_irreducible_iff_of_prime {p : ℕ} (hp : p.Prime) {a : K} :
+    Irreducible (X ^ p - C a) ↔ ∀ b, b ^ p ≠ a :=
+  ⟨(pow_ne_of_irreducible_X_pow_sub_C · dvd_rfl hp.ne_one), X_pow_sub_C_irreducible_of_prime hp⟩
+
+theorem X_pow_mul_sub_C_irreducible
+    {n m : ℕ} {a : K} (hm : Irreducible (X ^ m - C a))
+    (hn : ∀ (E : Type u) [Field E] [Algebra K E] (x : E) (hx : minpoly K x = X ^ m - C a),
```

Comment:
It's probably not very important, but `E` here needs to be in `Type u` or can this polymorphic?

Reply:
I think if you use [Fact (Irreducible (X ^ m - C a))] then K := AdjoinRoot K (X ^ m - C a) gets a Field instance, so you can just specialize to that.
(Are we re-reviewing [#9397](https://github.com/leanprover-community/mathlib4/commit/1b3fb343cf614b07bdaf8ab1f38429941b0006c3#diff-4a787b3a2a26263dcc69ab08fd5e84c0b54bd03c3d620c076236c6429f0dea7dR173-R177) now?)